lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,309 @@
+        if (data.unusedBytes) node.unusedBytes = (node.unusedBytes || 0) + data.unusedBytes;
+        if (data.duplicate !== undefined && isLastSegment) {
+          node.duplicate = data.duplicate;
+        }
+      });
+    }
+
+    const rootNode = newNode(sourceRoot);
+
+    // For every source file, apply the data to all components
+    // of the source path, creating nodes as necessary.
+    for (const [source, data] of Object.entries(sourcesData)) {
+      addNode(source || `<unmapped>`, data, rootNode);
+
+      // Apply the data to the rootNode.
+      rootNode.resourceBytes += data.resourceBytes;
+      if (data.unusedBytes) rootNode.unusedBytes = (rootNode.unusedBytes || 0) + data.unusedBytes;
+    }
+
+    /**
+     * Collapse nodes that have only one child.
+     * @param {Node} node
+     */
+    function collapse(node) {
+      while (node.children && node.children.length === 1) {
+        node.name += '/' + node.children[0].name;
+        node.children = node.children[0].children;
+      }
+
+      if (node.children) {
+        for (const child of node.children) {
+          collapse(child);
+        }
+      }
+    }
+    collapse(rootNode);
+
+    // TODO(cjamcl): Should this structure be flattened for space savings?
+    // Like DOM Snapshot.
+    // Less JSON (no super nested children, and no repeated property names).
+
+    return rootNode;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<RootNode[]>}
+   */
+  static async makeJavaScriptRootNodes(artifacts, context) {
+    /** @type {RootNode[]} */
+    const rootNodes = [];
+
+    const bundles = await JsBundles.request(artifacts, context);
+    const duplication = await ModuleDuplication.request(artifacts, context);
+
+    /** @type {Array<{src: string, length: number, unusedJavascriptSummary?: import('../computed/unused-javascript-summary.js').Summary}>} */
+    const scriptData = [];
+    const inlineScriptData = {
+      src: artifacts.URL.finalUrl,
+      length: 0,
+    };
+    for (const scriptElement of artifacts.ScriptElements) {
+      // Normalize ScriptElements so that inline scripts show up as a single entity.
+      if (!scriptElement.src) {
+        inlineScriptData.length += (scriptElement.content || '').length;
+        continue;
+      }
+
+      const url = scriptElement.src;
+      const bundle = bundles.find(bundle => url === bundle.script.src);
+      const scriptCoverages = artifacts.JsUsage[url];
+      if (!bundle || !scriptCoverages) continue;
+
+      scriptData.push({
+        src: scriptElement.src,
+        length: (scriptElement.content || '').length,
+        unusedJavascriptSummary:
+          await UnusedJavaScriptSummary.request({url, scriptCoverages, bundle}, context),
+      });
+    }
+    if (inlineScriptData.length) scriptData.unshift(inlineScriptData);
+
+    for (const {src, length, unusedJavascriptSummary} of scriptData) {
```

Comment:
this might be a little simpler to follow if it separated the inline/not-inline cases, something like

```js
let inlineScriptLength = 0;
for (const scriptElement of artifacts.ScriptElements) {
  // Normalize ScriptElements so that inline scripts show up as a single entity.
  if (!scriptElement.src) {
    inlineScriptLength += (scriptElement.content || '').length;
  }
}
const name = artifacts.URL.finalUrl;
if (inlineScriptLength) {
  rootNodes.push({
    name,
    node: {
      name,
      resourceBytes: inlineScriptLength,
      unusedBytes: 0,
      executionTime: 0,
    },
  });
}

const bundles = await JsBundles.request(artifacts, context);
const duplication = await ModuleDuplication.request(artifacts, context);

for (const scriptElement of artifacts.ScriptElements) {
  if (!scriptElement.src) {
    continue;
  }

  const src = scriptElement.src;
  const bundle = bundles.find(bundle => src === bundle.script.src);
  const scriptCoverages = artifacts.JsUsage[src];
  if (!bundle || !scriptCoverages) continue;

  const length = (scriptElement.content || '').length;
  const unusedJavascriptSummary = await UnusedJavaScriptSummary.request({url: src, scriptCoverages, bundle}, context);

  const name = src;
  // The rest of the existing second loop...
}
```

It would make it clearer exactly what happens to the inline script info, and it would get rid of the intermediate `scriptData` objects that are immediately destructured, the 2x find of the bundle (and no need to check it exists twice), etc.